File: PennMobile/src/main/java/com/pennapps/labs/pennmobile/BookGsrFragment.java
package com.pennapps.labs.pennmobile;


import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.pennapps.labs.pennmobile.api.Labs;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

import butterknife.Bind;
import butterknife.ButterKnife;
import retrofit.Callback;
import retrofit.RetrofitError;
import retrofit.client.Response;

public class BookGsrFragment extends Fragment {

    private Labs mLabs;

    @Bind(R.id.first_name)
    EditText firstName;
    @Bind(R.id.last_name) EditText lastName;
    @Bind(R.id.gsr_email) EditText email;
    @Bind(R.id.submit_gsr)
    Button submit;

    private String gsrID, gsrLocationCode, startTime, endTime;


    public BookGsrFragment() {
        // Required empty public constructor
    }

    public static BookGsrFragment newInstance(String gsrID, String gsrLocationCode, String startTime, String endTime) {
        BookGsrFragment fragment = new BookGsrFragment();
        Bundle args = new Bundle();
        args.putString("gsrID", gsrID);
        args.putString("gsrLocationCode", gsrLocationCode);
        args.putString("startTime", startTime);
        args.putString("endTime", endTime);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            gsrID = getArguments().getString("gsrID");
            gsrLocationCode = getArguments().getString("gsrLocationCode");
            startTime = getArguments().getString("startTime");
            endTime = getArguments().getString("endTime");
        }
        mLabs = MainActivity.getLabsInstance();
        getActivity().setTitle(R.string.gsr);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.gsr_details_book, container, false);
        ButterKnife.bind(this, v);
        submit.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                if (firstName.getText().toString().matches("") || lastName.getText().toString().matches("")
                        || email.getText().toString().matches("")) {
                    Toast.makeText(getActivity(), "Please fill in all fields before booking",
                            Toast.LENGTH_LONG).show();
                } else {
                    bookGSR(Integer.parseInt(gsrID), Integer.parseInt(gsrLocationCode), startTime, endTime);
                }
            }
        });
        return v;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        ButterKnife.unbind(this);
    }

    private void bookGSR(int gsrId, int gsrLocationCode, String startTime, String endTime){

        mLabs.bookGSR(

                //Passing the values
                gsrLocationCode,
                gsrId,
                startTime,
                endTime,
                firstName.getText().toString(),
                lastName.getText().toString(),
                email.getText().toString(),
                "Penn Mobile GSR",
                "2158986533",
                "2-3",

                //Creating an anonymous callback
                new Callback<Response>() {
                    @Override
                    public void success(Response result, Response response) {
                        //On success we will read the server's output using bufferedreader
                        //Creating a bufferedreader object
                        BufferedReader reader = null;

                        //An string to store output from the server
                        String output = "";

                        try {
                            //Initializing buffered reader
                            reader = new BufferedReader(new InputStreamReader(result.getBody().in()));

                            //Reading the output in the string
                            output = reader.readLine();
                        } catch (IOException e) {
                            e.printStackTrace();
                        }

                        //Displaying the output as a toast
                        Toast.makeText(getActivity(), "GSR successfully booked", Toast.LENGTH_LONG).show();
                    }

                    @Override
                    public void failure(RetrofitError error) {
                        //If any error occurred displaying the error as toast
                        Toast.makeText(getActivity(), "An error has occurred. Please try again." ,Toast.LENGTH_LONG).show();
                    }
                }
        );
    }

}

File: PennMobile/src/main/java/com/pennapps/labs/pennmobile/BookGsrFragment.kt
package com.pennapps.labs.pennmobile

import android.graphics.Color
import android.graphics.PorterDuff
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.preference.PreferenceManager
import com.pennapps.labs.pennmobile.api.StudentLife
import com.pennapps.labs.pennmobile.classes.GSRBookingResult
import com.pennapps.labs.pennmobile.databinding.GsrDetailsBookBinding
import retrofit.Callback
import retrofit.RetrofitError
import retrofit.client.Response

class BookGsrFragment : Fragment() {
    private var _binding: GsrDetailsBookBinding? = null
    val binding get() = _binding!!

    private lateinit var mActivity: MainActivity

    // fields for booking
    internal lateinit var firstNameEt: EditText
    internal lateinit var lastNameEt: EditText
    internal lateinit var emailEt: EditText

    // submit button
    private lateinit var submit: Button

    private lateinit var mStudentLife: StudentLife

    // gsr details
    private lateinit var gsrID: String
    private lateinit var gsrLocationCode: String
    private lateinit var startTime: String
    private lateinit var endTime: String
    private var gid: Int = 0
    private var roomId: Int = 0
    private lateinit var roomName: String

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        arguments?.let { arguments ->
            gsrID = arguments.getString("gsrID") ?: ""
            gsrLocationCode = arguments.getString("gsrLocationCode") ?: ""
            startTime = arguments.getString("startTime") ?: ""
            endTime = arguments.getString("endTime") ?: ""
            gid = arguments.getInt("gid")
            roomId = arguments.getInt("id")
            roomName = arguments.getString("roomName") ?: ""
        }
        mStudentLife = MainActivity.studentLifeInstance
        mActivity = activity as MainActivity
        mActivity.setTitle(R.string.gsr)
    }

    override fun onResume() {
        super.onResume()
        val mActivity: MainActivity = activity as MainActivity
        mActivity.setTitle(R.string.gsr)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        _binding = GsrDetailsBookBinding.inflate(inflater, container, false)
        val view = binding.root

        firstNameEt = binding.firstName
        lastNameEt = binding.lastName
        emailEt = binding.gsrEmail
        submit = binding.submitGsr

        // get user email and name from shared preferences if it's already saved
        val sp = PreferenceManager.getDefaultSharedPreferences(activity)
        val email = sp.getString(getString(R.string.email_address), "")
        val firstName = sp.getString(getString(R.string.first_name), "")
        val lastName = sp.getString(getString(R.string.last_name), "")

        firstNameEt.setText(firstName)
        lastNameEt.setText(lastName)
        emailEt.setText(email)

        submit.setOnClickListener {
            if (firstNameEt.text.toString().matches("".toRegex()) || lastNameEt.text.toString().matches("".toRegex()) ||
                emailEt.text.toString().matches("".toRegex())
            ) {
                Toast.makeText(
                    activity,
                    "Please fill in all fields before booking",
                    Toast.LENGTH_LONG,
                ).show()
            } else if (!emailEt.text.toString().matches("""\w+@(seas\.|sas\.|wharton\.|nursing\.)?upenn\.edu""".toRegex())) {
                Toast.makeText(activity, "Please enter a valid Penn email", Toast.LENGTH_LONG).show()
            } else {
                submit.isClickable = false
                submit.background.setColorFilter(Color.GRAY, PorterDuff.Mode.MULTIPLY)
                binding.loading.loadingPanel.visibility = View.VISIBLE
                bookGSR(Integer.parseInt(gsrID), gsrLocationCode, startTime, endTime, gid, roomId, roomName)
            }
        }

        mActivity.hideBottomBar()

        return view
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    private fun bookGSR(
        gsrId: Int,
        gsrLocationCode: String,
        startTime: String?,
        endTime: String?,
        gid: Int,
        roomId: Int,
        roomName: String,
    ) {
        mActivity.mNetworkManager.getAccessToken {
            var bearerToken = ""
            activity?.let { activity ->
                val sp = PreferenceManager.getDefaultSharedPreferences(activity)
                bearerToken = "Bearer " + sp.getString(getString(R.string.access_token), "").toString()
                Log.i("BookGSRFragment", bearerToken)
            }
            Log.i("BookGSRFragment", "Bearer $bearerToken")
            Log.i("BookGSRFragment", "Start $startTime")
            Log.i("BookGSRFragment", "End $endTime")
            Log.i("BookGSRFragment", "GID $gid")
            Log.i("BookGSRFragment", "ID $roomId")
            Log.i("BookGSRFragment", "Room Name $roomName")

            mStudentLife.bookGSR(
                // Passing the values
                bearerToken,
                startTime,
                endTime,
                gid,
                roomId,
                roomName,
                // Creating an anonymous callback
                object : Callback<GSRBookingResult> {
                    override fun success(
                        result: GSRBookingResult,
                        response: Response,
                    ) {
                        // Displaying the output as a toast and go back to GSR fragment
                        if (result.getDetail().equals("success")) {
                            Toast.makeText(activity, "GSR successfully booked", Toast.LENGTH_LONG).show()

                            // Save user info in shared preferences
                            val sp = PreferenceManager.getDefaultSharedPreferences(activity)
                            val editor = sp.edit()
                            editor.putString(getString(R.string.first_name), firstNameEt.text.toString())
                            editor.putString(getString(R.string.last_name), lastNameEt.text.toString())
                            editor.putString(getString(R.string.email_address), emailEt.text.toString())
                            editor.apply()
                        } else {
                            Toast.makeText(activity, "GSR booking failed", Toast.LENGTH_LONG).show()
                            Log.e("BookGsrFragment", "GSR booking failed with " + result.getError())
                        }
                        // go back to GSR fragment
                        binding.loading.loadingPanel.visibility = View.GONE
                        activity?.onBackPressed()
                    }

                    override fun failure(error: RetrofitError) {
                        // If any error occurred displaying the error as toast
                        Log.e("BookGSRFragment", "Error booking gsr", error)
                        Toast.makeText(activity, "An error has occurred. Please try again.", Toast.LENGTH_LONG).show()
                        binding.loading.loadingPanel.visibility = View.GONE
                        activity?.onBackPressed()
                    }
                },
            )
        }
    }

    companion object {
        fun newInstance(
            gsrID: String,
            gsrLocationCode: String,
            startTime: String,
            endTime: String,
            gid: Int,
            roomId: Int,
            roomName: String,
        ): BookGsrFragment {
            val fragment = BookGsrFragment()
            val args = Bundle()
            args.putString("gsrID", gsrID)
            args.putString("gsrLocationCode", gsrLocationCode)
            args.putString("startTime", startTime)
            args.putString("endTime", endTime)
            args.putInt("gid", gid)
            args.putInt("id", roomId)
            args.putString("roomName", roomName)
            fragment.arguments = args
            return fragment
        }
    }
}
